### PR TITLE
Feature/security

### LINKED ISSUE #11: Security
Creating security using JWT

### PR DESCRIPTION
**Pull request by wojciechkostecki**
Spring security with JWT.

Close #11

### CODE REVIEW

File: src/main/java/pl/wojciechkostecki/revel/controller/AuthController.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    public AuthController(RegisterService registerService, AuthenticationManager authenticationManager,
+                          JwtTokenUtil jwtTokenUtil, UserMapper userMapper) {
+        this.registerService = registerService;
+        this.authenticationManager = authenticationManager;
+        this.jwtTokenUtil = jwtTokenUtil;
+        this.userMapper = userMapper;
+    }
+
+    @PostMapping("/registration")
+    public ResponseEntity<Void> registrationUser(@Valid @RequestBody UserDTO userDTO) {
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/controller/AuthController.java]:
```suggestion
    public ResponseEntity<Void> registerUser(@Valid @RequestBody UserDTO userDTO) {
```

File: src/main/java/pl/wojciechkostecki/revel/controller/AuthController.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    public AuthController(RegisterService registerService, AuthenticationManager authenticationManager,
+                          JwtTokenUtil jwtTokenUtil, UserMapper userMapper) {
+        this.registerService = registerService;
+        this.authenticationManager = authenticationManager;
+        this.jwtTokenUtil = jwtTokenUtil;
+        this.userMapper = userMapper;
+    }
+
+    @PostMapping("/registration")
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/controller/AuthController.java]:
```suggestion
    @PostMapping("/register")
```

File: src/main/java/pl/wojciechkostecki/revel/controller/AuthController.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        try {
+            Authentication authenticate = authenticationManager
+                    .authenticate(new UsernamePasswordAuthenticationToken(request.getUsername(), request.getPassword()));
+
+            UserDetails user = (UserDetails) authenticate.getPrincipal();
+
+            return ResponseEntity.ok()
+                    .header(HttpHeaders.AUTHORIZATION, jwtTokenUtil.generateAccessToken(user))
+                    .body(userMapper.toDto(user));
+        } catch (BadCredentialsException ex) {
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/controller/AuthController.java]: Convert this exception with our global exception handler, remove this try catch

File: src/main/java/pl/wojciechkostecki/revel/model/User.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import javax.persistence.*;
+import java.util.HashSet;
+import java.util.Set;
+
+@Entity
+@Data
+public class User {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(name = "id")
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/model/User.java]: why do we even need this? It doesnt add anything - column will be named `id` by default

File: src/main/java/pl/wojciechkostecki/revel/model/dto/UserDTO.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import pl.wojciechkostecki.revel.model.UserRole;
+
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.Size;
+import java.util.HashSet;
+import java.util.Set;
+
+@Data
+public class UserDTO {
+
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/model/dto/UserDTO.java:13]: Worth adding id to this dto object?

File: src/main/java/pl/wojciechkostecki/revel/repository/UserRoleRepository.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package pl.wojciechkostecki.revel.repository;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+import pl.wojciechkostecki.revel.model.Role;
+import pl.wojciechkostecki.revel.model.UserRole;
+
+@Repository
+public interface UserRoleRepository extends JpaRepository<UserRole,Long> {
+    UserRole findByName(Role name);
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/repository/UserRoleRepository.java]: Probably better to return Optional here

File: src/main/java/pl/wojciechkostecki/revel/security/JwtTokenFilter.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Optional;
+
+@Component
+public class JwtTokenFilter extends OncePerRequestFilter {
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/security/JwtTokenFilter.java:21]: You have to understand how this filter works and what filter is in spring security if you will have this code in your project.

File: src/main/java/pl/wojciechkostecki/revel/security/JwtTokenUtil.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+@Component
+@RequiredArgsConstructor
+public class JwtTokenUtil {
+    private final String jwtSecret = "zdtlD3JK56m6wTTgsNFhqzjqP";
+    private final String jwtIssuer = "wojciechkostecki.pl";
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    public String generateAccessToken(UserDetails user) {
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/security/JwtTokenUtil.java:23]: Same here, what is jwt, how is it built, etc. Most of the stuff about JWT sine they might ask you about this on your interview.

File: src/main/java/pl/wojciechkostecki/revel/security/SecurityConfig.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        // Enable CORS and disable CSRF
+        http = http.cors().and().csrf().disable();
+
+        // Set session management to stateless
+        http = http
+                .sessionManagement()
+                .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+                .and();
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/security/SecurityConfig.java]: you dont need and here.

File: src/main/java/pl/wojciechkostecki/revel/security/SecurityConfig.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+        // Set unauthorized requests exception handler
+        http = http.exceptionHandling()
+                .authenticationEntryPoint
+                        ((request, response, ex) -> {response.sendError(HttpServletResponse.SC_UNAUTHORIZED, ex.getMessage());})
+                .and();
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/security/SecurityConfig.java]: same here .

File: src/main/java/pl/wojciechkostecki/revel/security/SecurityConfig.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        http = http.cors().and().csrf().disable();
+
+        // Set session management to stateless
+        http = http
+                .sessionManagement()
+                .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+                .and();
+
+        // Set unauthorized requests exception handler
+        http = http.exceptionHandling()
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/security/SecurityConfig.java]: What is this part doing?

File: src/main/java/pl/wojciechkostecki/revel/security/SecurityConfig.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        // Add JWT token filter
+        http.addFilterBefore(
+                jwtTokenFilter,
+                UsernamePasswordAuthenticationFilter.class
+        );
+    }
+
+    // Used by spring security if CORS is enabled.
+    @Bean
+    public CorsFilter corsFilter() {
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/security/SecurityConfig.java:52]: Read about CORS and what it is.

File: src/main/java/pl/wojciechkostecki/revel/service/RegisterService.java [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+    public RegisterService(UserRepository userRepository, UserMapper userMapper, PasswordEncoder passwordEncoder) {
+        this.userRepository = userRepository;
+        this.userMapper = userMapper;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    public void registerUser(UserDTO userDTO) {
+        if (userRepository.findByUsername(userDTO.getUsername()).isPresent()) {
+            throw new RuntimeException("There is a user with given login");
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/service/RegisterService.java]: Dont throw RuntimeException. Create custom one.

File: src/main/java/pl/wojciechkostecki/revel/model/dto/UserDTO.java (hunk 5 above)
Comment (wojciechkostecki) [reply src/main/java/pl/wojciechkostecki/revel/model/dto/UserDTO.java:13]: There was no such need at the moment, but I will add if it should be useful.

File: src/main/java/pl/wojciechkostecki/revel/security/JwtTokenUtil.java (hunk 8 above)
Comment (wojciechkostecki) [reply src/main/java/pl/wojciechkostecki/revel/security/JwtTokenUtil.java:23]: Ok, thanks for the hint.

File: src/main/java/pl/wojciechkostecki/revel/security/SecurityConfig.java (hunk 11 above)
Comment (wojciechkostecki) [reply src/main/java/pl/wojciechkostecki/revel/security/SecurityConfig.java]: This part was supposed to set up exception handling for unauthorized requests, but it is actually unnecessary and as you suggested earlier, I have removed it.

File: src/main/java/pl/wojciechkostecki/revel/service/RegisterService.java (hunk 13 above)
Comment (wojciechkostecki) [reply src/main/java/pl/wojciechkostecki/revel/service/RegisterService.java]: done

File: src/main/java/pl/wojciechkostecki/revel/exception/UserAlreadyExistException.java [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package pl.wojciechkostecki.revel.exception;
+
+public class UserAlreadyExistException extends RuntimeException{
```
Comment (rbobko93) [inline src/main/java/pl/wojciechkostecki/revel/exception/UserAlreadyExistException.java]: it could be LoginAlreadyUsedException

**Review by rbobko93 — APPROVED** [review]

**Pull request merged**